components/service/glean/src/main/java/mozilla/components/service/glean/storages/PingStorageEngine.kt

```diff
@@ -0,0 +1,180 @@
+            writePingToFile(pingFile, pingPath, pingData)
+        }
+    }
+
+    /**
+     * Function to deserialize and process all serialized ping files.  This function will ignore
+     * files that don't match the UUID regex and just delete them to prevent files from polluting
+     * the ping storage directory.
+     *
+     * @param processingCallback Callback function to do the actual process action on the ping.
+     *                           Typically this will be the [HttpPingUploader.upload] function.
+     * @return Boolean representing the success of the upload task. This may be the value bubbled up
+     *         from the callback, or if there was an error reading the files.
+     */
+    fun process(processingCallback: (String, String, Configuration) -> Boolean): Boolean {
+        var success = true
+
+        listPingFiles()?.forEach { file ->
+            if (file.name.matches(Regex(FILE_PATTERN))) {
+                if (!processFile(file, processingCallback)) {
+                    logger.error("Error processing ping file: ${file.name}")
+                    success = false
+                }
+            } else {
+                // Delete files that don't match the UUID FILE_PATTERN regex
+                file.delete()
+            }
+        }
+
+        return success
+    }
+
+    /**
+     * This function encapsulates processing of a single ping file.
+     *
+     * @param file The [File] to process
+     * @param processingCallback the callback that actually processes the file
+     *
+     */
+    private fun processFile(
+        file: File,
+        processingCallback: (String, String, Configuration) -> Boolean
+    ): Boolean {
+        BufferedReader(FileReader(file)).use {
+            try {
+                val path = it.readLine()
+                val serializedPing = it.readLine()
+
+                val processed = serializedPing == null ||
+                    processingCallback(path, serializedPing, Glean.configuration)
+
+                return if (processed) {
+                    file.delete()
+                    true
+                } else {
+                    // The callback couldn't process this file.
+                    false
+                }
+            } catch (e: FileNotFoundException) {
+                // This shouldn't happen after we queried the directory.
+                logger.error("Could not find ping file ${file.name}")
+                return false
+            } catch (e: IOException) {
+                // Something is not right.
+                logger.error("IO Exception when reading file ${file.name}")
+                return false
+            }
+        }
+    }
+
+    /**
+     * Helper function to list all serialized ping files or returns null if none
+     *
+     * @return An Array<File?> containing the list of ping files
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun listPingFiles(): Array<File>? {
```

Comment:
nit: does having this function buy us anything now? Can't we simply use `storageDirectory.lsitFiles()` at the call site?

Reply:
True